plugins/autoclean.c

```diff
@@ -75,58 +84,103 @@ static const char *datastore_path(const tal_t *ctx,
 		       subsystem_to_str(subsystem), field);
 }
 
-static struct command_result *set_next_timer(struct plugin *plugin)
-{
-	plugin_log(plugin, LOG_DBG, "setting next timer");
-	cleantimer = plugin_timer(plugin, time_from_sec(cycle_seconds), do_clean, plugin);
-	return timer_complete(plugin);
-}
-
-static struct command_result *clean_finished_one(struct command *cmd)
+static struct command_result *clean_finished(struct clean_info *cinfo)
 {
-	assert(cleanup_reqs_remaining != 0);
-	if (--cleanup_reqs_remaining > 0)
-		return command_still_pending(cmd);
-
 	for (enum subsystem i = 0; i < NUM_SUBSYSTEM; i++) {
-		if (num_cleaned[i] == 0)
+		if (!cinfo->num_cleaned[i])
 			continue;
 
-		jsonrpc_set_datastore_string(plugin, cmd,
+		plugin_log(plugin, LOG_DBG, "cleaned %"PRIu64" from %s",
+			   cinfo->num_cleaned[i], subsystem_to_str(i));
+		total_cleaned[i] += cinfo->num_cleaned[i];
+		jsonrpc_set_datastore_string(plugin, cinfo->cmd,
 					     datastore_path(tmpctx, i, "num"),
-					     tal_fmt(tmpctx, "%"PRIu64, num_cleaned[i]),
+					     tal_fmt(tmpctx, "%"PRIu64, total_cleaned[i]),
 					     "create-or-replace", NULL, NULL, NULL);
 
 	}
 
-	return set_next_timer(plugin);
+	/* autoclean-once? */
+	if (cinfo->cmd) {
```

Comment:
This is safe for races? I.e., if the user closes the connection with the command in flight, `cinfo->cmd` doesn't get reset, correct? Otherwise we might end up starting multiple timer loops.

Reply:
Yes, since autoclean-once allocates cinfo dynamically, each command will be a separate one (you might get lots of messages is they try to delete at the same time ofc!)